### ISSUE TITLE
Lua Scripting - Submiting via JS

### ISSUE THREAD

**Issue by AlexIzydorczyk** [scrapinghub/splash#200]
Hi,

I've been having trouble getting a lua script that calls something like:

```
splash:runjs("$(':submit')[0].click()")
```

to run. That is, splash doesn't seem to actually render the result of the page after submission - as if the submit button had been clicked in the first place.

Any idea how to work around this or what I may be doing wrong?

Thanks,
Alex

**Comment by microhello** [comment]
+1 
I've been encounter error when executing lua script like:
```Lua
function main(splash)   
    assert(splash:go(splash.args.url))
    splash:wait(0.5)       
    splash:evaljs("document.getElementById('a').click();")   
    splash:wait(0.5)  
end
```
error message is:
```
JsError(u"TypeError: 'undefined' is not a function",)
```

**Comment by jjmurre** [comment]
I am strugging wlth exactly the same problem. The dom element is available, however, it does not have a 'click()' method. Also, pre-loading jquery using splash:autoload does not help.

**Comment by kmike** [comment]
@AlexIzydorczyk I think the problem is that splash:runjs doesn't wait until the redirect is done. So this script shows the main webpage:

```
splash:go("http://yandex.ru")
splash:runjs("$(':submit')[0].click()")
return splash:png()
```

But this script prints a page after form submission:

```
splash:go("http://yandex.ru")
splash:runjs("$(':submit')[0].click()")
splash:wait(1)
return splash:png()
```

@microhello Are you sure document.getElementById('a') returns an HTMLElement which has a 'click' method? Not all HTML elements have .click() method, and there are some rules regarding event bubbling (https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/click). Also note that `getElementById('a')` finds an element with `id='a'`, not necessarily `<a href='">` element. Splash itself doesn't provide click() method, it is a standard browser feature. It is better to check it in a browser console first.

@jjmurre do you have an example? I wonder why isn't splash:autoload working. JQuery doesn't provide this method on HTMLElements, so `document.getElementById('a').click()` won't work with jQuery if it didn't work before; JQuery provides $(elem).click() method.

**Comment by jjmurre** [comment]
I am having the following spider:

    import scrapy

    SPLASH_SCRIPT = """
    function main(splash)
        assert(splash:go(splash.args.url))
        ok, reason = splash:runjs("document.getElementsByTagName('a')[0].click()")
        splash:wait(1)
        return {ok=ok, reason=reason}
    end
    """


    class ExampleSpider(scrapy.Spider):
        name = 'example'
        start_urls = ['http://172.17.42.1:8000']  # IP of localhost

        def start_requests(self):
            for url in self.start_urls:
                yield scrapy.Request(url, self.parse, meta={
                    'splash': {
                        'endpoint': 'execute',
                        'args': {'lua_source': SPLASH_SCRIPT}
                    }
                })

        def parse(self, response):
            print response.body

And I am serving the following html at http://localhost:8000/

  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <title>title</title>
    </head>
    <body>
      <a href="./target.html" id="link">Link</a>
      <button id="btn">Button</button>
    </body>
  </html>

The response from scrapy is:

2015-06-16 21:11:32+0200 [scrapy] INFO: Scrapy 0.24.6 started (bot: profiler)
2015-06-16 21:11:32+0200 [scrapy] INFO: Optional features available: ssl, http11
2015-06-16 21:11:32+0200 [scrapy] INFO: Overridden settings: {'NEWSPIDER_MODULE': 'profiler_scrapy.spiders', 'SPIDER_MODULES': ['profiler_scrapy.spiders'], 'DUPEFILTER_CLASS': 'scrapyjs.SplashAwareDupeFilter', 'HTTPCACHE_STORAGE': 'scrapyjs.SplashAwareFSCacheStorage', 'BOT_NAME': 'profiler'}
2015-06-16 21:11:32+0200 [scrapy] INFO: Enabled extensions: LogStats, TelnetConsole, CloseSpider, WebService, CoreStats, SpiderState
2015-06-16 21:11:32+0200 [scrapy] INFO: Enabled downloader middlewares: HttpAuthMiddleware, DownloadTimeoutMiddleware, UserAgentMiddleware, RetryMiddleware, DefaultHeadersMiddleware, MetaRefreshMiddleware, HttpCompressionMiddleware, RedirectMiddleware, CookiesMiddleware, SplashMiddleware, ChunkedTransferMiddleware, DownloaderStats
2015-06-16 21:11:32+0200 [scrapy] INFO: Enabled spider middlewares: HttpErrorMiddleware, OffsiteMiddleware, RefererMiddleware, UrlLengthMiddleware, DepthMiddleware
2015-06-16 21:11:32+0200 [scrapy] INFO: Enabled item pipelines: SaveResults
2015-06-16 21:11:32+0200 [example] INFO: Spider opened
2015-06-16 21:11:32+0200 [example] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2015-06-16 21:11:32+0200 [scrapy] DEBUG: Telnet console listening on 127.0.0.1:6023
2015-06-16 21:11:32+0200 [scrapy] DEBUG: Web service listening on 127.0.0.1:6080
2015-06-16 21:11:33+0200 [example] DEBUG: Crawled (200) <POST http://localhost:8051/execute> (referer: None)
{"reason": "TypeError: 'undefined' is not a function"}
2015-06-16 21:11:33+0200 [example] INFO: Closing spider (finished)
2015-06-16 21:11:33+0200 [example] INFO: Dumping Scrapy stats:
        {'downloader/request_bytes': 507,
         'downloader/request_count': 1,
         'downloader/request_method_count/POST': 1,
         'downloader/response_bytes': 169,
         'downloader/response_count': 1,
         'downloader/response_status_count/200': 1,
         'finish_reason': 'finished',
         'finish_time': datetime.datetime(2015, 6, 16, 19, 11, 33, 563916),
         'log_count/DEBUG': 3,
         'log_count/INFO': 7,
         'response_received_count': 1,
         'scheduler/dequeued': 2,
         'scheduler/dequeued/memory': 2,
         'scheduler/enqueued': 2,
         'scheduler/enqueued/memory': 2,
         'splash/execute/request_count': 1,
         'splash/execute/response_count/200': 1,
         'start_time': datetime.datetime(2015, 6, 16, 19, 11, 32, 535539)}
2015-06-16 21:11:33+0200 [example] INFO: Spider closed (finished)

**Comment by kmike** [comment]
Aha, you're right. There are two issues. 

1. WebKit version Splash uses doesn't provide .click() function for `<a>` elements - only buttons (`<input>` elements) work. This may be fixed by https://github.com/scrapinghub/splash/pull/222, but I haven't checked it.

2. jQuery `.click()` method doesn't allow to follow links - see e.g. http://stackoverflow.com/questions/1694595/can-i-call-jquery-click-to-follow-an-a-link-if-i-havent-bound-an-event-hand.

A workaround (until Splash get a nice library to do these tasks) is to use `window.location = document.getElementById('link').href`:

```lua
function main(splash)
  assert(splash:go(splash.args.url))
  splash:runjs("window.location = document.getElementsByTagName('a')[0].href")
  splash:wait(0.5)
  return {
    html = splash:html(),
    png = splash:png(),
    har = splash:har(),
  }
end
```

Another workaround is to use a library other than jQuery. For example, one can use CasperJS [clientutils](http://docs.casperjs.org/en/latest/modules/clientutils.html):

```lua
function main(splash)
  -- preload CasperJS client utils.
  -- __utils__ object is compatible with CasperJS
  splash:autoload("https://raw.githubusercontent.com/n1k0/casperjs/master/modules/clientutils.js")
  splash:autoload([[
  	window.__utils__ = new ClientUtils({});
  ]])
  
  -- go to a page and click a link 
  assert(splash:go(splash.args.url))
  assert(splash:runjs("__utils__.click('#link')"))
  splash:wait(0.5)
  return {
    html = splash:html(),
    png = splash:png(),
    har = splash:har(),
  }
end
```

 As an example, I'm loading CasperJS ClientUtils from github; for production it is better to find a CDN URL for autoload or serve it from your own server.

**Comment by jjmurre** [comment]
@kmike : That explains things. Thanx for your great answer! I will keep an eye on splash updates and use the approaches you are describing in the meantime.

**Comment by pauloram** [comment]
@kmike: Thanks for this, awesome stuff, works well to get additional content to load that depends on clicking a "Load more" button. I can't figure out how to make this iterative though so that it keeps 'clicking' on the Load More button until whilst it continues to appear. Any ideas?

**Comment by kmike** [comment]
Hey @pauloram,

A simplest way to do it is to write a loop:

```lua
for i = 1,10 do
    click_load_more()
    splash:wait(1.0)
end
```

You can also stop the loop earlier if you detect that the content is not appearing; the logic is likely website-specific.

**Comment by pauloram** [comment]
Thanks @kmike that makes sense but the challenge I am having is making the CasperJS click ( assert(splash:runjs("__utils__.click('#link')"))) conditional on the presence of #link. I thought I could use __utils__.exists() but now I am not so sure...

I only want to return the HTML once I have clicked on "Load more" all the times required for the page to load up all the additional content.

**Comment by Youwotma** [comment]
You can test the existence of the element and click it while it exists like this:
```
while splash:evaljs('!!document.querySelector("#link")') do
    assert(splash:runjs("utils.click('#link')")))
end
```

**Comment by linusred** [comment]
Came across this issue today while looking into splash for the first time. Same problem above, solutions worked well, but I ended up doing something like this: 
    //fill in a form, submit form via javascript
    while (string.find(splash:evaljs("$('elementToTestOnNextPage').text()"), "Foo") == nil) do
    splash:wait(0.5)
  end

As well, which seemed to work alright. Although having an API to wait for the request would really be great.

**Comment by kmike** [comment]
> Came across this issue today while looking into splash for the first time. Same problem above, solutions worked well, but I ended up doing something like this

There is a very basic `wait_for` function example in docs (http://splash.readthedocs.org/en/1.8/scripting-tutorial.html#writing-modules) - if we want to bundle such function it could make sense to at least add timeouts / max iteration count:
```lua
function wait_for(condition)
    while not condition() do
        splash:wait(0.05)
    end
end
```

Your example can be abstracted e.g. like this (still not perfect):
```lua
get_text = splash:jsfunc("function(sel) { return $(sel).text(); }") -- requires jquery

function wait_for_text(sel, text)
    wait_for(function() return string.find(get_text(sel), text) ~= nil end) 
end

-- usage
wait_for_text("elementToTestOnNextPage", "Foo")
```

Something similar can be done for requests (I haven't tested it):

```lua
function wait_for_request(condition)
    local request_found = false
    splash:on_request(request)
        if request_found then return end -- splash has no way to unregister a callback yet
        if condition(request) then request_found = true end
    end
    wait_for(function() return found end)
end

-- usage
wait_for_request(function(request)
    return request.url == "my_url"
end)
```

You get the idea.

> As well, which seemed to work alright. Although having an API to wait for the request would really be great.

@linusred how can this API look like? Maybe you can post a code snippet with the desired API?

**Comment by linusred** [comment]
Thanks @kmike , that definitely helps. Hadn't noticed the module section of the api.

Personally, I think it'd benefit from a simple API that allows splash to wait for a specific resource or URL change before preforming further actions. 

splash:waitForResource("image.png")

Or

splash:waitForUrl:("http://foo.com/?blarg='foo'")

Or

splash:waitForSelector('foo.bar') --When foo exists

I'm sure these could all be abstracted out to modules currently, as per your example, but I think they're common enough amongst all use cases that they belong in the default API.

**Comment by kmike** [comment]
Yeah, I agree, we need a built-in library for common tasks.

**Comment by psychok7** [comment]
Hey guys, i am using `Splash 2.0.2 + Scrapy 1.0.5` and im still not able to render javascript with a click. Here is an example url https://olx.pt/anuncio/loja-nova-com-250m2-garagem-em-box-fechada-para-arrumos-IDyTzAT.html#c49d3d94cf

```
class OlxSpider(scrapy.Spider):
    name = "olx"
    rotate_user_agent = True
    allowed_domains = ["olx.pt"]
    start_urls = [
        "https://olx.pt/imoveis/"
    ]

    def parse(self, response):
        script = """
        function main(splash)
            splash:go(splash.args.url)
            splash:runjs('document.getElementById("contact_methods").getElementsByTagName("span")[1].click();')
            splash:wait(0.5)
            return splash:html()
        end
        """
        for href in response.css('.link.linkWithHash.detailsLink::attr(href)'):
            url = response.urljoin(href.extract())
            yield scrapy.Request(url, callback=self.parse_house_contents, meta={
                'splash': {
                    'args': {'lua_source': script},
                    'endpoint': 'execute',
                }
            })

        for next_page in response.css('.pager .br3.brc8::attr(href)'):
            url = response.urljoin(next_page.extract())
            yield scrapy.Request(url, self.parse)

    def parse_house_contents(self, response):

        import ipdb;ipdb.set_trace()
```

how can i get this to work?

**Comment by kmike** [comment]
There are now scraping helpers which allow to select elements, submit forms, etc. Fixed by #490 and #531.
